### PR TITLE
Umpire logs refactor - table output + adding percentages

### LINKED ISSUE #2304: Update memory statistics table
**What is the requested feature?**
Add minimum and average (across ranks) memory utilization statistics from umpire and update format of the statistics table.

**Is your request related to a specific problem?**
No

**Describe the solution you'd like**
Here is an example of what we currently have:
```
Umpire          DEVICE sum across ranks: 15293.4 GB
Umpire          DEVICE         rank max:   31.3 GB
Umpire       DEVICE::0 sum across ranks: 15293.4 G
Umpire       DEVICE::0         rank max:   31.3 GB
Umpire            HOST sum across ranks: 7272.2 GB
Umpire            HOST         rank max:   14.7 GB
Umpire HYPRE_DEVICE_POOL sum across ranks: 7178.4 GB
Umpire HYPRE_DEVICE_POOL         rank max:   14.8 GB
Umpire          PINNED sum across ranks:  256.3 GB
Umpire          PINNED         rank max:  589.6 MB
```

This is what I propose:
```
+-------------------+---------------+---------------+---------------+--------------+
|       Umpire Pool |    Min (GB/%) |    Max (GB/%) |    Avg (GB/%) |   Sum (TB/%) |
|                   |    over ranks |    over ranks |      per rank |    all ranks |
+-------------------+---------------+---------------+---------------+--------------+
|            DEVICE |   31.3  (49%) |   31.3  (49%) |   31.3  (49%) |  15.2  (24%) |
|              HOST |   14.7 (2.9%) |   14.7 (2.9%) |   14.7 (2.9%) |   7.2 (2.9%) |
| HYPRE_DEVICE_POOL |   14.8  (23%) |   14.8  (23%) |   14.8  (23%) |   7.1  (13%) |
|            PINNED |  0.589 (0.1%) |  0.589 (0.1%) |  0.589 (0.1%) |   0.3 (0.1%) |
+-------------------+---------------+---------------+---------------+--------------+

```
**Describe alternatives you've considered**
N/A

**Additional context**
Having access to minimum, maximum, and average utilization leads to a better understanding of load balancing.

### PR DESCRIPTION
**Pull request by arng40**
#### What we currently have :

```
Umpire          DEVICE sum across ranks: 15293.4 GB
Umpire          DEVICE         rank max:   31.3 GB
Umpire       DEVICE::0 sum across ranks: 15293.4 G
Umpire       DEVICE::0         rank max:   31.3 GB
Umpire            HOST sum across ranks: 7272.2 GB
Umpire            HOST         rank max:   14.7 GB
Umpire HYPRE_DEVICE_POOL sum across ranks: 7178.4 GB
Umpire HYPRE_DEVICE_POOL         rank max:   14.8 GB
Umpire          PINNED sum across ranks:  256.3 GB
Umpire          PINNED         rank max:  589.6 MB
```

#### This is what I propose inspired by the display proposed by @victorapm 
```  
+-----------------------------+---------------------+---------------------+---------------------+--------------------+
|         Umpire Memory Pool  |     Min over ranks  |    Max  over ranks  |    Avg  over ranks  |    Sum over ranks  |
|  (reserved / % over total)  |                     |                     |                     |                    |
+-----------------------------+---------------------+---------------------+---------------------+--------------------+
|                     DEVICE  |    1.4 MB   (0.0%)  |    1.7 MB   (0.0%)  |    1.5 MB   (0.0%)  |   6.1 MB   (0.0%)  |
|                  DEVICE::0  |    1.4 MB   (0.0%)  |    1.7 MB   (0.0%)  |    1.5 MB   (0.0%)  |   6.1 MB   (0.0%)  |
|                       HOST  |    2.3 MB   (0.0%)  |    2.5 MB   (0.0%)  |    2.4 MB   (0.0%)  |   9.8 MB   (0.0%)  |
|              HYPRE_UM_POOL  |  504.7 KB   (0.0%)  |  645.8 KB   (0.0%)  |  542.4 KB   (0.0%)  |   2.1 MB   (0.0%)  |
|                     PINNED  |  488.1 KB   (0.0%)  |  963.0 KB   (0.0%)  |  725.5 KB   (0.0%)  |   2.8 MB   (0.0%)  |
|                         UM  |    4.0 GB  (25.3%)  |    4.0 GB  (25.3%)  |    4.0 GB  (25.3%)  |  16.0 GB  (25.3%)  |
+-----------------------------+---------------------+---------------------+---------------------+--------------------+
```
#### Code side
- I've also refactored a bit the `addUmpireHighWaterMarks()` function to modern C++

### CODE REVIEW

File: src/coreComponents/common/initializeEnvironment.cpp [hunk 1]
```diff
@@ -291,16 +295,36 @@ static void addUmpireHighWaterMarks()
...
+                                    ( MpiWrapper::sum( mark ) / MpiWrapper::commSize() )  * 100 );
+    string sumMarkValue = GEOS_FMT( "{} ({}%)",
+                                    LvArray::system::calculateSize( MpiWrapper::sum( mark ) ),
+                                    ( MpiWrapper::sum( mark ) / MpiWrapper::sum( mark ) ) * 100 );
+
+    tableData.addRow( allocatorName,
+                      minMarkValue,
+                      maxMarkValue,
+                      avgMarkValue,
+                      sumMarkValue );
```
Comment (TotoGaz) [inline src/coreComponents/common/initializeEnvironment.cpp]: First, add the `const` keyword whenever you can.
Second, every time you call `MpiWrapper::{min, max, sum}`, you make a collective exchange. In the previous code, there was one call for each data. Please refactor your code not to multiply the call for no added value.

File: src/coreComponents/common/initializeEnvironment.cpp [hunk 2]
```diff
@@ -291,16 +295,36 @@ static void addUmpireHighWaterMarks()
...
 
     pushStatsIntoAdiak( allocatorName + " sum across ranks", mark );
     pushStatsIntoAdiak( allocatorName + " rank max", mark );
   }
+
+  TableLayout memoryStatLayout ( {"Umpire Pool", "Min (GB/%)\nover ranks", "Max (GB/%)\nover ranks", "Avg (GB/%)\nover ranks", "Sum (GB/%)\nover ranks" } );
+  TableTextFormatter memoryStatLog( memoryStatLayout );
```
Comment (TotoGaz) [inline src/coreComponents/common/initializeEnvironment.cpp]: You're surely able to add `const` as well.

File: src/coreComponents/mesh/generators/WellGeneratorBase.cpp [hunk 3]
```diff
@@ -141,9 +144,10 @@ void WellGeneratorBase::generateWellGeometry( )
   // make sure that the perforation locations are valid
   checkPerforationLocationsValidity();
 
-  if( getLogLevel() >= 1 )
+  if( getLogLevel() >= 1 && logger::internal::rank == 0 )
```
Comment (TotoGaz) [inline src/coreComponents/mesh/generators/WellGeneratorBase.cpp]: The `internal` may clue you into guessing it's meant for internal purposes 🤣 😉 🤗 
Please use `MpiWrapper::comRank()` instead.

**Review by TotoGaz — COMMENTED** [review]
I'd double check that the `umpire` layout is not used by others for performance checks.

File: src/coreComponents/mesh/generators/WellGeneratorBase.cpp (hunk 3 above)
Comment (MelReyCG) [reply src/coreComponents/mesh/generators/WellGeneratorBase.cpp]: Hi @TotoGaz, about these `logger::internal`, I'm going to refactor them all by hiding them in a `Logger` class and provide an assessor (we discussed of that quite a long time ago). There are like 30 of them currently ;)

Comment (TotoGaz) [reply src/coreComponents/mesh/generators/WellGeneratorBase.cpp]: No pb 😉

**Comment by paveltomin** [comment]
@arng40 @MelReyCG any update?

**Comment by arng40** [comment]
> @arng40 @MelReyCG any update?

**Comment by arng40** [comment]
@paveltomin 
This PR is being tested by a user on a simulation to see whether the results are consistent or not

**Comment by arng40** [comment]
@paveltomin umpire values are consistent, you can have a look to the structure and let me know if it's good

**Comment by paveltomin** [comment]
> @paveltomin umpire values are consistent, you can have a look to the structure and let me know if it's good

looks good, thanks

**Review by victorapm — APPROVED** [review]
Nice work! Thanks for implementing this

**Comment by rrsettgast** [comment]
@arng40 Is there something inside of umpire to do this? It might make sense to do this sort of change within umpire for all to use.

**Comment by CusiniM** [comment]
this PR has broken my mac builds. Looks like that `_SC_AVPHYS_PAGES` is not defined on macOS.

**Comment by herve-gross** [comment]
@CusiniM : try to replace by `_SC_PHYS_PAGES`

**Comment by CusiniM** [comment]
> @CusiniM : try to replace by `_SC_PHYS_PAGES`

but are they the same thing? I have patched it like this: https://github.com/GEOS-DEV/GEOS/pull/3207

**Pull request merged**